modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+
+        self.in_chans = in_channels
+        self.out_chans = out_channels
+        self.inp_shape = inp_shape
+        self.patch_size = patch_size
+        self.num_features = self.embed_dim = embed_dim
+        self.num_blocks = num_blocks
+        self.modulate_filter = modulate_filter
+        self.modulate_mlp = modulate_mlp
+        self.scale_shift_mode = scale_shift_mode
+        norm_layer = partial(nn.LayerNorm, eps=1e-6)
+
+        self.patch_embed = PatchEmbed(
+            inp_shape=inp_shape,
+            in_channels=self.in_chans,
+            patch_size=self.patch_size,
+            embed_dim=embed_dim,
+        )
+        num_patches = self.patch_embed.num_patches
+
+        self.pos_embed = nn.Parameter(torch.zeros(1, num_patches, embed_dim))
+        self.pos_drop = nn.Dropout(p=drop_rate)
+
+        self.h = inp_shape[0] // self.patch_size[0]
+        self.w = inp_shape[1] // self.patch_size[1]
+
+        self.blocks = nn.ModuleList(
+            [
+                Block(
+                    embed_dim=embed_dim,
+                    mod_dim=mod_dim,
+                    num_blocks=self.num_blocks,
+                    mlp_ratio=mlp_ratio,
+                    drop=drop_rate,
+                    norm_layer=norm_layer,
+                    sparsity_threshold=sparsity_threshold,
+                    hard_thresholding_fraction=hard_thresholding_fraction,
+                    modulate_filter=modulate_filter,
+                    modulate_mlp=modulate_mlp,
+                    scale_shift_mode=scale_shift_mode,
+                )
+                for i in range(depth)
+            ]
+        )
+
+        self.head = nn.Linear(
+            embed_dim,
+            self.out_chans * self.patch_size[0] * self.patch_size[1],
+            bias=False,
+        )
+
+        torch.nn.init.trunc_normal_(self.pos_embed, std=0.02)
+        self.apply(self._init_weights)
+
+        self.mod_additive_proj = nn.Linear(mod_dim, embed_dim)
+        if not (modulate_mlp or modulate_filter):
+            self.mod_embed_net = nn.Identity()
+        else:
+            embed_model = {} if embed_model is None else embed_model
+            self.mod_embed_net = ModEmbedNet(**embed_model)
+
+    def _init_weights(self, m: nn.Module):
+        """Init model weights"""
+        if isinstance(m, nn.Linear):
+            torch.nn.init.trunc_normal_(m.weight, std=0.02)
+            if isinstance(m, nn.Linear) and m.bias is not None:
+                nn.init.constant_(m.bias, 0)
+        elif isinstance(m, nn.LayerNorm):
+            nn.init.constant_(m.bias, 0)
+            nn.init.constant_(m.weight, 1.0)
+
+    def forward_features(self, x: Tensor, mod: Tensor) -> Tensor:
+        """Forward pass of core ModAFNO"""
+        B = x.shape[0]
+        x = self.patch_embed(x)
+        x = x + self.pos_embed
+        x = self.pos_drop(x)
+
+        mod_embed = self.mod_embed_net(mod)
+        mod_additive = self.mod_additive_proj(mod_embed).unsqueeze(dim=(1))
```

Comment:
Just Confirm that dim=1 is always valid and that mod will never be higher dimensional needing something like dim=-1c

Reply:
Yes, here `dim==1` should be always valid. The shape of `x` here is `(B, P, C)` where the `P` contains all the AFNO patches flattened into one dimension. So we create a singleton dimension in `mod_additive` to make its shape `(B, 1, C)` to make all patches get the same modulation. The flattening to a `(B, P, C)` shape should always happen (even if this happens to be turned in to a 3D ModAFNO later).